Improve type definitions for the viewer

ex37 opened the pull request:
This commit improves compatibility of the viewer code with TypeScript by including missing type imports/definitions and correcting existing ones

Snuffleupagus commented:
Given that the default viewer is *not* intended to be directly used by third-parties, and thus isn't included in `pdfjs-dist`, I don't know if this is really desirable. Please see http://mozilla.github.io/pdf.js/getting_started/#introduction (emphasis mine):
> The viewer is built on the display layer and is the UI for PDF viewer in Firefox and the other browser extensions within the project. It can be a good starting point for building your own viewer. ***However, we do ask if you plan to embed the viewer in your own site, that it not just be an unmodified version. Please re-skin it or build upon it.***

ex37 commented:
@Snuffleupagus this could help people who want to extend the viewer. Plus there are also lot's of nice utils within the web folder. Also added jsdoc types could be helpful for project's contributors.

But if you think this is not within the scope of the project, we can remove the entrypoint, and just add the type fixes.

Snuffleupagus commented:
> But if you think this is not within the scope of the project, we can remove the entrypoint, and just add the type fixes.

That'd be a requirement as far as I'm concerned. Please note that the existing type definitions are unused within the PDF.js project, but has nonetheless added a bunch of work for the core contributors; hence we don't want to further increase that.

Also, please make sure to write a good *commit message*.

Snuffleupagus commented on web/annotation_editor_params.js:
```diff
@@ -15,6 +15,20 @@
 
 import { AnnotationEditorParamsType } from "pdfjs-lib";
 
+/** @typedef {import("./event_utils.js").EventBus} EventBus */
```
Here, and everywhere else throughout the patch, imported typedefs should be placed *before* the regular `import`s.

Snuffleupagus commented on web/base_tree_viewer.js:
```diff
@@ -43,14 +43,14 @@ class BaseTreeViewer {
 
   /**
-   * @private
+   * @protected
```
Here, and everywhere else throughout the patch, why do we need to make this change?

ex37 replied:
cause typescript wise you can't redefine a private property in a child class (was leading to a broken .d.ts)

Snuffleupagus commented on web/base_tree_viewer.js:
```diff
@@ -71,7 +71,9 @@ class BaseTreeViewer {
    * Prepend a button before a tree item which allows the user to collapse or
    * expand all tree items at that level; see `_toggleTreeItem`.
-   * @private
+   * @protected
+   * @param {Element} div
+   * @param {any} [hidden=false]
```
```suggestion
   * @param {HTMLDivElement} div
   * @param {boolean} [hidden]
   * @protected
```

ex37 replied:
this method's signature is different for child classes (they are passing an object as a second parameter). In typescript this is not allowed, that's why I set hidden to any as a workaround.

Snuffleupagus replied:
No, they are not:
 - https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_outline_viewer.js#L185
 - https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_layer_viewer.js#L103

Hence the changes I mentioned are correct here for *the base* method.

ex37 replied:
the call to parent class is correct, but I am talking about the child classes' redefined method signature
https://github.com/mozilla/pdf.js/blob/d8331e31005edba6afef28ff2f4d55ffe5731df4/web/pdf_outline_viewer.js#L167

ex37 replied:
here is a simple demo, just for case:

https://www.typescriptlang.org/play?#code/MYGwhgzhAEAKYCcCmA7ALtA3gKGtAtkmgBYD2AJgBSIDmAXNCgK74BGSCAlFtrtAL7ZB2UJBgBhYgEsQ5aEgAeaVORjxk6HnkIkK1BPWitSpEEjApuOPHj6D+QA

ex37 commented on tsconfig.json:
```diff
@@ -16,8 +16,27 @@
       "fluent-bundle": ["./node_modules/@fluent/bundle/esm/index.js"],
       "fluent-dom": ["./node_modules/@fluent/dom/esm/index.js"],
-      "web-null_l10n": ["../web/genericl10n.js"]
+
```
@Snuffleupagus should I get rid of path aliases as well? Those have no use besides generating types, but was hard to figure out the list so could be helpful for someone else like myself.

Snuffleupagus replied:
Yes, given that we don't want to add unused code.

(Those shouldn't need to be figured out, since they can basically be copied from `gulpfile.mjs`.)

ex37 committed:
Improve type definitions for the viewer

This commit improves compatibility of the viewer code with TypeScript by including missing type imports/definitions and correcting existing ones

ex37 reviewed (commented):
Addressed requested changes, changed commit message / pr name.
Also added `eslint-disable-next-line` where needed.

Snuffleupagus commented on web/annotation_editor_params.js:
```diff
@@ -13,8 +13,22 @@
+ * @property {HTMLInputElement} editorInkColor
+ * @property {HTMLInputElement} editorInkThickness
+ * @property {HTMLInputElement} editorInkOpacity
+ * @property {HTMLElement} editorStampAddImage
```
```suggestion
 * @property {HTMLButtonElement} editorStampAddImage
```

Snuffleupagus commented on web/annotation_editor_params.js:
```diff
@@ -13,8 +13,22 @@
+ * @property {HTMLInputElement} editorInkOpacity
+ * @property {HTMLElement} editorStampAddImage
+ * @property {HTMLInputElement} editorFreeHighlightThickness
+ * @property {HTMLElement} editorHighlightShowAll
```
```suggestion
 * @property {HTMLButtonElement} editorHighlightShowAll
```

Snuffleupagus commented on web/external_services.js:
```diff
@@ -13,6 +13,8 @@
  * limitations under the License.
  */
 
+/** @typedef {import("./l10n.js").L10n} L10n */
```
```suggestion
/** @typedef {import("./interfaces.js").IL10n} IL10n */
```

Snuffleupagus commented on web/external_services.js:
```diff
@@ -28,6 +30,9 @@ class BaseExternalServices {
   reportTelemetry(data) {}
 
+  /**
+   * @returns {Promise<L10n>}
```
```suggestion
   * @returns {Promise<IL10n>}
```

Snuffleupagus commented on web/pdf_attachment_viewer.js:
```diff
@@ -27,6 +31,7 @@ import { waitOnEventOrTimeout } from "./event_utils.js";
 /**
  * @typedef {Object} PDFAttachmentViewerRenderParameters
  * @property {Object|null} attachments - A lookup table of attachment objects.
+ * @property {boolean} keepRenderedCapability
```
```suggestion
 * @property {boolean} [keepRenderedCapability]
```

Snuffleupagus commented on web/pdf_layer_viewer.js:
```diff
@@ -13,6 +13,11 @@
+/** @typedef {import("./event_utils.js").EventBus} EventBus */
+// eslint-disable-next-line max-len
+/** @typedef {import("../src/optional_content_config.js").OptionalContentConfig} OptionalContentConfig */
+/** @typedef {import("../src/document.js").PDFDocument} PDFDocument */
```
That file doesn't even exist, hence this cannot possibly work.
```suggestion
/** @typedef {import("../src/display/api").PDFDocumentProxy} PDFDocumentProxy
```

Snuffleupagus commented on web/pdf_outline_viewer.js:
```diff
@@ -13,6 +13,12 @@
+// eslint-disable-next-line max-len
+/** @typedef {import("./download_manager.js").DownloadManager} DownloadManager */
+/** @typedef {import("./interfaces").IPDFLinkService} IPDFLinkService */
+/** @typedef {import("../src/document.js").PDFDocument} PDFDocument */
```
```suggestion
/** @typedef {import("../src/display/api").PDFDocumentProxy} PDFDocumentProxy
```

Snuffleupagus commented:
https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

ex37 commented:
squashed

Snuffleupagus commented on web/grab_to_pan.js:
```diff
@@ -20,7 +20,7 @@ const CSS_CLASS_GRAB = "grab-to-pan-grab";
   /**
    * Construct a GrabToPan instance for a given HTML element.
-   * @param {Element} options.element
+   * @param {{element: Element}}
```
That's not a pattern we use anywhere in the code-base really.
It seems more correct to, just above, add
```javascript
/**
 * @typedef {Object} GrabToPanOptions
 * @property {HTMLElement} element
 */
```
and then write this as follows
```suggestion
   * @param {GrabToPanOptions} options
```

timvandermeij commented:
Before we can merge this please also include a better commit message. The guidelines at https://cbea.ms/git-commit/ help with this. The commit message should primarily show which change is being made and why in e.g. `git log` output; implementation details like which files or classes are changed can usually be left out because that's already clear from looking at the commit diff. A suggestion could be:

```
Improve type definitions for the viewer

This commit improves compatibility of the viewer code with TypeScript by
including missing type imports/definitions and correcting existing ones.
```

ex37 commented:
Got it. Set to your suggested description, as it fully describes what this pr is about.

timvandermeij commented on web/pdf_outline_viewer.js:
```diff
@@ -13,6 +13,13 @@
+/** @typedef {import("./event_utils.js").EventBus} EventBus */
+// eslint-disable-next-line max-len
+/** @typedef {import("./download_manager.js").DownloadManager} DownloadManager */
+/** @typedef {import("./interfaces").IPDFLinkService} IPDFLinkService */
```
Missing `.js` extension here as well.

timvandermeij reviewed (approved):
Looks good to me with the comments below addressed, but given the original review I'll leave the final approval to @Snuffleupagus. Thanks!

ex37 commented on web/pdf_document_properties.js:
```diff
@@ -13,6 +13,11 @@
+/** @typedef {import("./event_utils.js").EventBus} EventBus */
+/** @typedef {import("./interfaces.js").IL10n} IL10n */
+/** @typedef {import("./overlay_manager.js").OverlayManager} OverlayManager */
+/** @typedef {import("../src/display/api").PDFDocumentProxy} PDFDocumentProxy */
```
it works for typescript, yes, but for consistency makes sense to add extensions everywhere

ex37 commented on web/base_tree_viewer.js:
```diff
@@ -71,7 +71,9 @@ class BaseTreeViewer {
    * expand all tree items at that level; see `_toggleTreeItem`.
-   * @private
+   * @param {HTMLDivElement} div
+   * @param {boolean|object} [hidden=false]
```
it's required, as discussed here https://github.com/mozilla/pdf.js/pull/17879#discussion_r1549981654

timvandermeij replied:
Huh, that's interesting and something we might want to change later, but indeed not in this PR. Let's keep the object part of this comment as-is then.

ex37 commented:
removed the default value for `hidden`, and also added missing extensions

ex37 commented:
@timvandermeij in my opinion a proper fix for `_addToggleButton` method would probably be separating it into two methods, which will be using same signature across inheriting classes. Other than type definitions, in my understanding, it can also help with runtime optimizations.  But yes, that's outside of this pr scope.

Snuffleupagus reviewed (approved):
r=me, thanks.
